### ISSUE TITLE
Common Parameters should include a CancellationToken

### ISSUE THREAD

**Issue by rhubarb-geek-nz** [PowerShell/PowerShell#19685]
### Summary of the new feature / enhancement

There have been a number of requests to add timeouts to various commands ( why always in seconds ? ) it would be simpler to add [CancellationToken](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-7.0) support to [Common Parameters](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_commonparameters?view=powershell-7.3) so that all commands automatically have the capability in a consistent manner.

While superficially a timeout might seem obvious and simple, a CancellationToken is far more useful and allows the client control over when the cancellation is requested, eg not just because of a timeout but because some other condition occurred after the command has been started.

### Proposed technical implementation details (optional)

Proof of concept

This demonstrates a cancellation token working with Invoke-Command

Implementation - [CommandWithCancellationToken.cs](https://github.com/rhubarb-geek-nz/CancellationTokenEvent/blob/main/CancellationTokenEvent/CommandWithCancellationToken.cs)

Unit tests - [CommandWithCancellationTokenTests.cs](https://github.com/rhubarb-geek-nz/CancellationTokenEvent/blob/main/UnitTests/CommandWithCancellationTokenTests.cs)

**Comment by rhubarb-geek-nz** [comment]
The effect of the cancellation token should be inherited; 

- If a PowerShell is invoked and cancelled due to the cancellation token then all child PowerShells should be stopped/cancelled
- If a command invokes Jobs or ThreadJobs each of those jobs should inherit the cancellation. So even though the Start-Job/Start-ThreadJob has completed, the created jobs will be cancelled when the cancellation token is cancelled.
- In the case of parallel ForEach-Object, each parallel job should also be cancelled.
- If a new cancellation token is defined for a new scope, the original cancellation token still applies, so the effective is cumulative, eg if the first cancellation token expires after 10 minutes and 2 minutes in a task starts with a new cancellation token that expires after 20 minutes, the new task will be cancelled after 8 minutes.
- If a remote task is started, that should still be cancelled according to the current cancellation token.
- Any process started with Start-Process should be terminated on cancellation.

Perhaps the CancellationToken should be in a well-known variable like $ErrorActionPreference, eg as $CancellationToken.

**Comment by rhubarb-geek-nz** [comment]
The cancellation mechanism is not intended to give any hint regarding how long to go before the cancellation occurs. A cancellation due to timeout is only one method, other cancellations may be caused by the user pressing control-C, or parallel search requests invoked and the first successful response cancels the uncompleted redundant requests.

**Comment by rhubarb-geek-nz** [comment]
Existing Cmdlets should not need to change as the existing [Cmdlet.StopProcessing](https://learn.microsoft.com/en-us/dotnet/api/System.Management.Automation.Cmdlet.StopProcessing?view=powershellsdk-7.2.0) mechanism should be used the cancel running cmdlets.

**Comment by kasini3000** [comment]
> While superficially a timeout might seem obvious and simple, a CancellationToken is far more useful and allows the client control over when the cancellation is requested

Both of these are required.
**I reiterate that it is important for multithreading. Powershell is different from shell, as it works at the multi-threaded level.**

https://github.com/PowerShell/ThreadJob/issues/26

**Comment by rhubarb-geek-nz** [comment]
> Both of these are required.

This request is not about adding timeouts because a timeout can be implemented using a CancellationToken with  [CancelAfter](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.cancelafter?view=net-7.0) , the reverse is not true. Once cancellation tokens are implemented then you have the ability to do timeouts.

**Comment by rhubarb-geek-nz** [comment]
It would be nice to be triaged.

**Comment by rhubarb-geek-nz** [comment]
[GitHub Proof of concept](https://github.com/rhubarb-geek-nz/CancellationTokenEvent)
[PowerShell gallery demonstration](https://www.powershellgallery.com/packages/rhubarb-geek-nz.CancellationTokenEvent/1.0.2)

**Comment by Zetanova** [comment]
I found a method to get the SIGINT signal (windows and linux) from the pwsh.exe to be useable for sync-over-async calls on Runspace1
https://gist.github.com/Zetanova/9e1a0d5e35d9ce876840b9d9b22445b3

**Comment by rhubarb-geek-nz** [comment]
> I found a method to get the SIGINT signal (windows and linux) from the pwsh.exe to be useable for sync-over-async calls on gist [signal-cancellation](https://gist.github.com/Zetanova/9e1a0d5e35d9ce876840b9d9b22445b3)

I don't think that signals should be used in PowerShell programming. A script or cmdlet should stop according to either PowerShell.Stop or because a cancellation token has been signalled.

Signals should be managed by the top level process on the main thread and their scope is the entire process. A script or cmdlet should not care if it is the upper most script or a deeply nested script, run remotely, in a job or thread. It should have simple access to a cancellation token that tells it when it abort.

**Comment by Zetanova** [comment]
Yes, until the powershell team implements something they dont want to implement for years.
This SignalToken can be used to abort scripts in the pwsh.exe on script creation time.

The other option is off to kill pwsh.exe and open it again.
